### PR TITLE
[TS] Templating/data binding support (renderer + designer)

### PR DESCRIPTION
**Pull request by dclaux**
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/2995)

### CODE REVIEW

File: source/nodejs/package-lock.json [hunk 1]
```diff
@@ -497,14 +497,14 @@
       }
     },
     "@lerna/add": {
-      "version": "3.13.3",
-      "resolved": "https://registry.npmjs.org/@lerna/add/-/add-3.13.3.tgz",
-      "integrity": "sha512-T3/Lsbo9ZFq+vL3ssaHxA8oKikZAPTJTGFe4CRuQgWCDd/M61+51jeWsngdaHpwzSSRDRjxg8fJTG10y10pnfA==",
+      "version": "3.14.0",
```
Comment (shalinijoshi19) [inline source/nodejs/package-lock.json]: Curious -- upgrades to these packages driven by functionality need or security flagging or just because?

File: source/nodejs/README.md [hunk 2]
```diff
@@ -29,10 +30,10 @@ We use lerna to manage package linking and building.
...
 
 1. `cd adaptivecards-site`
-2. `npm install hexo-cli -g`
-3. `hexo generate`
-4. `hexo server`
-5. `Open up the browser to point to localhost:[portnumber] printed after the command above returns under "Hexo is running at".`
+2. `npx lerna run release`
+3. `npm install hexo-cli -g`
+5. `hexo server`
+6. Open up the browser to point to `localhost:[portnumber]` printed after the command above returns under "Hexo is running at".
```
Comment (shalinijoshi19) [inline source/nodejs/README.md]: Typo? Missing the backtick?

File: source/nodejs/adaptivecards/example.html [hunk 3]
```diff
@@ -53,7 +53,6 @@
 			adaptiveCard.hostConfig = new AdaptiveCards.HostConfig({
 				fontFamily: "Segoe UI, Helvetica Neue, sans-serif"
 				// More host config options
-				// You can find example configs here: https://github.com/microsoft/AdaptiveCards/tree/master/samples/HostConfig
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards/example.html]: curious -- why?

File: source/nodejs/README.md (hunk 2 above)
Comment (dclaux) [reply source/nodejs/README.md]: Not sure what you mean. This is markdown syntax, and I didn't write this.

File: source/nodejs/adaptivecards/example.html (hunk 3 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/example.html]: Sorry, why what?

File: source/nodejs/package-lock.json (hunk 1 above)
Comment (dclaux) [reply source/nodejs/package-lock.json]: These get updated as you compile. I have nothing to do with this change, and never understood why/when they happen.

File: source/nodejs/adaptivecards/example.html (hunk 3 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/example.html]: Guessing you're asking why the more informative line was removed - I didn't remove it. It's most likely due to automatic conflict resolution.

File: source/nodejs/README.md (hunk 2 above)
Comment (shalinijoshi19) [reply source/nodejs/README.md]: Yes but it's changed and am trying to explain the changes - Looks like these files i am commenting on that you dont recognize the changes to should be reverted 

---
In reply to: [297809373](https://github.com/microsoft/AdaptiveCards/pull/2995#discussion_r297809373) [](ancestors = 297809373)

File: source/nodejs/adaptivecards/example.html (hunk 3 above)
Comment (shalinijoshi19) [reply source/nodejs/adaptivecards/example.html]: yup - got it - then let's revert this file before the merge; 

---
In reply to: [297811427](https://github.com/microsoft/AdaptiveCards/pull/2995#discussion_r297811427) [](ancestors = 297811427)

File: source/nodejs/adaptivecards/package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "adaptivecards",
-	"version": "1.2.0",
+	"version": "1.3.0-alpha1",
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards/package.json]: >: "1.3.0-alpha1", [](start = 10, length = 17)

am guessing this change is by design and you made it?

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 5]
```diff
@@ -166,7 +166,14 @@ export abstract class SerializableObject {
     }
 
     setCustomProperty(name: string, value: any) {
-        this._rawProperties[name] = value;
+        let deleteProperty = (typeof value === "string" && Utils.isNullOrEmpty(value)) || value === undefined || value === null;
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards/src/card-elements.ts]: nit nit: shouldDeleteProperty might be more intuituve of a name here for the condition variable

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 6]
```diff
@@ -286,7 +293,18 @@ export abstract class CardElement extends CardObject {
             },
             this.separatorOrientation);
 
-        return renderedSeparator;
+            if (AdaptiveCard.alwaysBleedSeparators && renderedSeparator && this.separatorOrientation == Enums.Orientation.Horizontal) {
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards/src/card-elements.ts]: Nitnit: quick comment here to explain what this is condition is implying?

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 7]
```diff
@@ -491,8 +509,8 @@ export abstract class CardElement extends CardObject {
         Utils.setProperty(result, "separator", this.separator, false);
         Utils.setProperty(result, "height", this.height, "auto");
 
-        if (this.minPixelHeight) {
-            Utils.setProperty(result, "minHeight", this.minPixelHeight + "px");
+        if (this.supportsMinHeight) {
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards/src/card-elements.ts]: nit: this is a great variable name- very intuitive to read and understand. thanks!

**Comment by shalinijoshi19** [comment]
>             this._errorMessageElement.className = this.hostConfig.makeCssClassName("ac-input-validation-error-message");

Curious: Are these names at some point supposed to be localized?

---
Refers to: source/nodejs/adaptivecards/src/card-elements.ts:2833 in 30792ab. [](commit_id = 30792ab8c8de7bb69da9a72eb1a2ab3f0a061583, deletion_comment = False)

File: source/nodejs/adaptivecards-controls/src/popupmenu.ts [hunk 8]
```diff
@@ -100,7 +100,6 @@ export class PopupMenu extends PopupControl {
             this._overlayElement = document.createElement("div");
             this._overlayElement.className = "ms-ctrl-overlay";
             this._overlayElement.tabIndex = 0;
-            this._overlayElement.style.zIndex = "1000";
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards-controls/src/popupmenu.ts]: is this related to teh templating changes or bad merge?

File: source/nodejs/adaptivecards-controls/src/adaptivecards-controls.css [hunk 9]
```diff
@@ -1,363 +1,450 @@
...
-  0% {
-    opacity: 0; } }
+    0% {
+        opacity: 0;
+    }
+}
 
 @keyframes slideTopToBottom {
-  0% {
-    transform: translate(0, -20px); } }
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards-controls/src/adaptivecards-controls.css]: only spacing changes here looks like right?

File: source/nodejs/adaptivecards/package.json (hunk 4 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/package.json]: I don't see a reason why it shouldn't be 1.3.0-alpha1? It can always become anything else we want if necessary.

File: source/nodejs/adaptivecards/src/card-elements.ts (hunk 5 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts]: Yes, that's a very small nit :) Hopefully you don't mind if I don't fix.

File: source/nodejs/adaptivecards-controls/src/popupmenu.ts (hunk 8 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-controls/src/popupmenu.ts]: The popup menu control was too aggressive in wanting to be in the foreground, and interfered with the designer.

File: source/nodejs/adaptivecards-controls/src/adaptivecards-controls.css (hunk 9 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-controls/src/adaptivecards-controls.css]: I am not seeing this in my local file. What I have is correct. Don't know why lines 11 and 12 are marked as removed, they shouldn't be:
image

**Comment by dclaux** [comment]
> > ```
> >         this._errorMessageElement.className = this.hostConfig.makeCssClassName("ac-input-validation-error-message");
> > ```
> 
> Curious: Are these names at some point supposed to be localized?
> 
> Refers to: source/nodejs/adaptivecards/src/card-elements.ts:2833 in [30792ab](https://github.com/microsoft/AdaptiveCards/commit/30792ab8c8de7bb69da9a72eb1a2ab3f0a061583). [](commit_id = [30792ab](https://github.com/microsoft/AdaptiveCards/commit/30792ab8c8de7bb69da9a72eb1a2ab3f0a061583), deletion_comment = False)

CSS class names should never be localized.

File: source/nodejs/adaptivecards-controls/src/adaptivecards-controls.css (hunk 9 above)
Comment (shalinijoshi19) [reply source/nodejs/adaptivecards-controls/src/adaptivecards-controls.css]: Thanks good to know -- let me check. whitespace changes here.. formatting looks like? 

---
In reply to: [297859608](https://github.com/microsoft/AdaptiveCards/pull/2995#discussion_r297859608) [](ancestors = 297859608)

File: source/nodejs/adaptivecards/src/card-elements.ts (hunk 6 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts]: Added a comment.

File: source/nodejs/adaptivecards-controls/src/adaptivecards-controls.css (hunk 9 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-controls/src/adaptivecards-controls.css]: Maybe.

File: source/nodejs/adaptivecards/package.json (hunk 4 above)
Comment (shalinijoshi19) [reply source/nodejs/adaptivecards/package.json]: I am not sure why this is getting changed esp since this ends pu in a totatlly separate package right? This shouldnt be touched wiht this change yes?

---
In reply to: [297856584](https://github.com/microsoft/AdaptiveCards/pull/2995#discussion_r297856584) [](ancestors = 297856584)

File: source/nodejs/adaptivecards/example.html (hunk 3 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/example.html]: I re-added that comment.

File: source/nodejs/README.md (hunk 2 above)
Comment (dclaux) [reply source/nodejs/README.md]: I honestly don't know.

File: source/nodejs/adaptivecards-designer/index.html [hunk 10]
```diff
@@ -18,7 +18,9 @@
 
 	<script type="text/javascript">
 		window.onload = function() {
-			if (!ACDesigner.SettingsManager.isLocalStorageAvailable) {
+            ACDesigner.GlobalSettings.previewFeaturesEnabled = true;
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards-designer/index.html]: >ACDesigner.GlobalSettings.previewFeaturesEnabled = true; [](start = 12, length = 56)

Do we want this to be enabled by default in masteR?

File: source/nodejs/adaptivecards-designer/package-lock.json [hunk 11]
```diff
@@ -229,6 +229,11 @@
 				"acorn": "^5.0.0"
 			}
 		},
+		"adaptivecards": {
+			"version": "1.2.0",
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards-designer/package-lock.json]: > "1.2.0", [](start = 13, length = 9)

You are continuing to use 1.2.0 here but then you end up changing the version for the package.json adaptivecards package to a different value --or is there no correlation?

File: source/nodejs/adaptivecards-designer/index.html (hunk 10 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-designer/index.html]: Good catch - no, we don't want that. I turned it off.

File: source/nodejs/adaptivecards-designer/package-lock.json (hunk 11 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-designer/package-lock.json]: I have nothing to do with any `package-lock.json` file. They are automatically generated and I don't know how. I also don't know why this one references 1.2.

File: source/nodejs/adaptivecards/package.json (hunk 4 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/package.json]: Yes it should. We have shipped 1.2, what lives in master needs to have a newer version.

**Comment by shalinijoshi19** [comment]
Cool thanks! 

---
In reply to: [506038582](https://github.com/microsoft/AdaptiveCards/pull/2995#issuecomment-506038582) [](ancestors = 506038582)

File: source/nodejs/adaptivecards-designer-app/samples/Restaurant.data.json [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+		"addressRegion": "WA",
+		"streetAddress": "16851 Redmond Way",
+		"postalCode": "98052",
+		"addressCountry": "US"
+	},
+	"image": "https://s3-media1.fl.yelpcdn.com/bphoto/HD_NsxwaCTwKRxvOZs2Shw/ls.jpg",
+	"telephone": "+14258816461",
+	"aggregateRating": {
+		"reviewCount": 176,
+		"@type": "AggregateRating",
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards-designer-app/samples/Restaurant.data.json]: INteresting -- are we saying we may have an alternate way with databinding to implement rating controls?

File: source/nodejs/adaptivecards-designer-app/src/app.ts [hunk 13]
```diff
@@ -10,6 +11,9 @@ import "adaptivecards-designer/lib/adaptivecards-designer.css";
 import "adaptivecards-designer/node_modules/adaptivecards-controls/lib/adaptivecards-controls.css";
 
 window.onload = function() {
+    // Comment to disable preview features (data binding)
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards-designer-app/src/app.ts]: does this comment still apply ?

File: source/nodejs/adaptivecards-templating/src/expression-parser.ts [hunk 14]
```diff
@@ -0,0 +1,772 @@
...
+                    i += matches[0].length;
+
+                    matchFound = true;
+
+                    break;
+                }
+            }
+
+            if (!matchFound) {
+                throw new Error("Unexpected character " + subExpression[0] + " at position " + i);
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards-templating/src/expression-parser.ts]: >Error [](start = 26, length = 5)

Nit: maybe we should end up using our own custom error class here if one doesnt already exist instead of generic Error

File: source/nodejs/adaptivecards-designer-app/samples/Restaurant.data.json (hunk 12 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-designer-app/samples/Restaurant.data.json]: No, this is just data. To display that data as a "rating" you need to have some sort of a rating control whether it's ChoiceSet or an explicit Rating control doesn't matter)

File: source/nodejs/adaptivecards-designer-app/src/app.ts (hunk 13 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-designer-app/src/app.ts]: I updated the comment.

File: source/nodejs/adaptivecards-templating/src/expression-parser.ts (hunk 14 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-templating/src/expression-parser.ts]: Derived error classes are nonstandard in JS/TS. There is no way to catch a specific error type as far as I know, one has to write something like this (assuming RangeError extends Error):
```typescript
try {
    throw new RangeError();
}
catch (e) {
    if(e instanceof RangeError){
        console.log('out of range');
    }
}
```
So essentially you always throw a generic Error anyway.

File: source/nodejs/adaptivecards-designer-app/samples/Restaurant.data.json (hunk 12 above)
Comment (shalinijoshi19) [reply source/nodejs/adaptivecards-designer-app/samples/Restaurant.data.json]: got it! tx

---
In reply to: [297899768](https://github.com/microsoft/AdaptiveCards/pull/2995#discussion_r297899768) [](ancestors = 297899768)

File: source/nodejs/adaptivecards-templating/src/expression-parser.ts (hunk 14 above)
Comment (shalinijoshi19) [reply source/nodejs/adaptivecards-templating/src/expression-parser.ts]: Hmm TS isnt unique in that error handling pattern -- other languages also enable custom error handling code that looks like that.. So that doesnt seem justifcation to not use custom errors in the code base; Infact preferring custom error types does seem to be a best practice of sorts and not that unheard of in TS/JS -- found this --https://blog.joefallon.net/2018/09/typescript-try-catch-finally-and-custom-errors/ and it lists precisely this as a best practice - "Best Practice – Be Specific with Errors
Often you will want to throw or return errors that are more specific than the generic Error object. That is easy to do by extending the standard error object. The goal with having a more specific error is to maximize the chance that the error can be handled locally and allow code execution to continue. Here is how you catch the specific errors."  I think we should revisit this pattern in our code base.. can probably sync up a bit offline as well as something we can work on together later?

---
In reply to: [297901240](https://github.com/microsoft/AdaptiveCards/pull/2995#discussion_r297901240) [](ancestors = 297901240)
